### PR TITLE
Partial projections pushdown

### PR DESCRIPTION
**Pull request by phd3**
Splitting partial projection extraction from #1720. cc @martint @wagnermarkd 

Some notes:

Currently, PushProjectionIntoTableScan is applied only in cases where all expressions are completely convertible to a connector level expressions. The only expressions that satisfy this criteria are simple column references, dereference expressions and constants. Any expression that involves anything more complex will not invoke PushProjectionIntoTableScan. So there is no opportunity for pushing down such expressions into the connector.

For example, for projections [“a.b.c”, “d”], the PushProjectionIntoTableScan rule will be invoked. But, for projections [“a.b.c” + “c.d”, “e”], it will be a no-op. This is very primitive, because such a pushdown will not work for even slightly complex projections. In such cases, we know that even if “a.b.c” + “c.d” cannot be pushed down, the individual subexpressions “a.b.c” and “c.d” can be. We need to implement the partial projection pushdown support for that. 

After the partial projection support, the PushProjectionIntoTableScan would look like the following:

Blank Diagram (5)

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/spi/expression/ConnectorExpressionTranslator.java [hunk 1]
```diff
@@ -82,7 +82,7 @@ public Expression translate(ConnectorExpression expression)
             if (expression instanceof FieldDereference) {
                 FieldDereference dereference = (FieldDereference) expression;
 
-                RowType type = (RowType) expression.getType();
+                RowType type = (RowType) dereference.getTarget().getType();
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/spi/expression/ConnectorExpressionTranslator.java]: is it possible to test this?

also, cmt title is too long. Maybe just `Fix translation of FieldDereference to DereferenceExpression`.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushProjectionIntoTableScan.java [hunk 2]
```diff
@@ -102,6 +103,9 @@ public Result apply(ProjectNode project, Captures captures, Context context)
             return Result.empty();
         }
 
+        List<ConnectorExpression> newConnectorProjections = result.get().getProjections();
+        checkState(newConnectorProjections.size() == projections.size(), "mismatch between input and output projections from the connector");
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushProjectionIntoTableScan.java]: Capitalize "Mismatch", include list sizes in the error message.

```suggestion
        checkState(newConnectorProjections.size() == projections.size(), "Mismatch between input and output projections from the connector: expected %s but got %s", projections.size(), newConnectorProjections.size());
```

(maybe you need to rewrap this)

File: presto-main/src/main/java/io/prestosql/spi/expression/ConnectorExpressionTranslator.java [hunk 3]
```diff
@@ -52,13 +53,13 @@ public static Expression translate(ConnectorExpression expression, Map<String, S
...
 
-    public static ConnectorExpression translate(Session session, Expression expression, TypeAnalyzer types, TypeProvider inputTypes)
+    public static Optional<ConnectorExpression> translate(Session session, Expression expression, TypeAnalyzer types, TypeProvider inputTypes)
     {
         return new SqlToConnectorExpressionTranslator(types.getTypes(session, inputTypes, expression))
                 .process(expression);
     }
 
-    private static class ConnectorToSqlExpressionTranslator
+    static class ConnectorToSqlExpressionTranslator
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/spi/expression/ConnectorExpressionTranslator.java]: why not `private`? please revert

File: presto-main/src/main/java/io/prestosql/spi/expression/ConnectorExpressionUtil.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        private final ConnectorExpressionTranslator.SqlToConnectorExpressionTranslator translator;
+
+        PartialTranslator(Map<NodeRef<Expression>, Type> types, ImmutableMap.Builder<NodeRef<?>, ConnectorExpression> translatedSubExpressions)
+        {
+            requireNonNull(types, "types is null");
+            this.translatedSubExpressions = requireNonNull(translatedSubExpressions, "translatedSubExpressions is null");
+            this.translator = new ConnectorExpressionTranslator.SqlToConnectorExpressionTranslator(types);
+        }
+
+        public Void visitExpression(Expression node, Void context)
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/spi/expression/ConnectorExpressionUtil.java]: `@Override`

File: presto-main/src/main/java/io/prestosql/spi/expression/ConnectorExpressionUtil.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    /**
+     * Traverse an expression recursively to replace subexpressions with an alternative expression, if specified in the
+     * {@param replacements} map.
+     */
+    public static Expression replaceNodeRefs(Expression expression, Map<NodeRef<?>, Expression> replacements)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new NodeRefReplacer(replacements), expression);
+    }
+
+    private static class NodeRefReplacer
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/spi/expression/ConnectorExpressionUtil.java]: We have a very similar class for this already: `ReferenceAwareExpressionNodeInliner`

File: presto-main/src/main/java/io/prestosql/spi/expression/ConnectorExpressionUtil.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+import io.prestosql.sql.tree.ExpressionRewriter;
+import io.prestosql.sql.tree.ExpressionTreeRewriter;
+import io.prestosql.sql.tree.NodeRef;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class ConnectorExpressionUtil
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/spi/expression/ConnectorExpressionUtil.java]: I generally try to avoid generic "Util" classes, as they tend to become junk drawers over time. Instead, I'd create dedicated utilities (i.e., `PartialTranslator`, `ExpressionReplacer`). See, for example `ExpressionNodeInliner`, `ExpressionExtractor` or `SortExpressionExtractor`.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushProjectionIntoTableScan.java [hunk 7]
```diff
@@ -72,39 +77,40 @@ public Result apply(ProjectNode project, Captures captures, Context context)
...
+                partialProjectionsBuilder.add(expr);
+            });
         }
 
-        Map<String, ColumnHandle> assignments = tableScan.getAssignments()
-                .entrySet().stream()
+        List<NodeRef<?>> nodesForPartialProjections = nodeReferencesBuilder.build();
+        List<ConnectorExpression> connectorPartialProjections = partialProjectionsBuilder.build();
+
+        Map<String, ColumnHandle> assignments = tableScan.getAssignments().entrySet().stream()
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushProjectionIntoTableScan.java]: Unnecessary formatting change. This was previously:

```java
        Map<String, ColumnHandle> assignments = tableScan.getAssignments()
                .entrySet().stream()
```

File: presto-main/src/main/java/io/prestosql/spi/expression/ConnectorExpressionTranslator.java (hunk 1 above)
Comment (phd3) [reply presto-main/src/main/java/io/prestosql/spi/expression/ConnectorExpressionTranslator.java]: @findepi Fixed the message. Added a test in `TestConnectorExpressionTranslator`. I haven't added literals for the test since the translation for them would rely on how ExpressionAnalyzer and LiteralEncoder.

**Comment by phd3** [comment]
@martint Thanks for the review, I've addressed your comments.

**Comment by findepi** [comment]
@phd3 please see Travis.

**Comment by phd3** [comment]
thanks @findepi, I've fixed related errors.

File: presto-main/src/main/java/io/prestosql/sql/planner/ReferenceAwareExpressionNodeInliner.java [hunk 8]
```diff
@@ -33,7 +33,7 @@ public static Expression replaceExpression(Expression expression, Map<NodeRef<Ex
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/ReferenceAwareExpressionNodeInliner.java]: Rephrase the commit message: "Update constructor visibility in ReferenceAwareExpressionNodeInliner"

**Comment by phd3** [comment]
@martint Modified the commit message and cleanly rebased.

**Comment by phd3** [comment]
@martint There's one non-trivial change for catering to lambda expressions that I've made after the rebase. All tests seem to pass now. Curious if there are other cases of context-dependent symbols like in lambda expressions.

**Comment by martint** [comment]
> There's one non-trivial change for catering to lambda expressions that I've made after the rebase. All tests seem to pass now. Curious if there are other cases of context-dependent symbols like in lambda expressions.

The only other case would be in the context of a correlated subquery. But those appear at the "top level" from the point of view of the plan operation (i.e., not within a nested context like lambda expressions), are are not bound to columns.

**Comment by martint** [comment]
Merged, thanks for working on this!
